### PR TITLE
[Console] Fix question formatting using SymfonyStyle::ask()

### PR DESCRIPTION
**Pull request by chalasr**
| Q             | A
| ------------- | ---
| Branch?       | 2.7
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #20964 
| License       | MIT
| Doc PR        | n/a

Given
```php
$io = new SymfonyStyle($input, $output);
$io->ask('Do you want to use Foo\\Bar <comment>or</comment> Foo\\Baz\\?', 'Foo\\Bar');
```

Before output
before

After output
after

### CODE REVIEW

**Comment by ogizanagi** [comment]
👍 

Status: Reviewed

File: src/Symfony/Component/Console/Tests/Helper/SymfonyQuestionHelperTest.php [hunk 1]
```diff
@@ -101,6 +103,15 @@ public function testAskEscapeLabel()
         $this->assertOutputContains('Do you want a \?', $output);
```
Comment (ogizanagi) [inline src/Symfony/Component/Console/Tests/Helper/SymfonyQuestionHelperTest.php]: This test does not longer look relevant, and anyway I think it didn't covered the [initial issue](https://github.com/symfony/symfony/issues/19493) 😕 

Changing it for the following:

```php

    public function testAskEscapeLabel()
    {
        $helper = new SymfonyQuestionHelper();
        $helper->setInputStream($this->getInputStream('sure'));
        $helper->ask($this->createInputInterfaceMock(), $output = $this->createOutputInterface(), new Question('Add a \\'));

        $this->assertOutputContains('Add a \\', $output);
    }
```

will result into the following failure:

```php
 1) Symfony\Component\Console\Tests\Helper\SymfonyQuestionHelperTest::testAskEscapeLabel
Failed asserting that ' Add a </info>:
 > ' contains "Add a \".
```

Looks edgy to have a simple backslash at the end, but it actually was the initial issue. It seems there is no need to escape anything else 😕

Comment (chalasr) [inline src/Symfony/Component/Console/Tests/Helper/SymfonyQuestionHelperTest.php]: The initial issue was about having a backslash as default value which has been fixed, not really about a question ending with a `\`, which is indeed a very edge case to me. I think this is a good consensus, don't you?

Comment (ogizanagi) [inline src/Symfony/Component/Console/Tests/Helper/SymfonyQuestionHelperTest.php]: As I said, your fix is legitimate, but there is still another issue elsewhere.

That said, the test mentioned does not actually cover anything I think and should be removed ^^'

Comment (chalasr) [inline src/Symfony/Component/Console/Tests/Helper/SymfonyQuestionHelperTest.php]: Thanks for looking at it. The useless test has been removed

**Comment by sstok** [comment]
So to be clear, if you have a value with backslashes like a namespace you need to replace these before usage?

**Comment by ogizanagi** [comment]
@sstok: You'll at least need to escape the last character if it's a backslash using `OutputFormatter::escape()`, or replace it by `<<`.

Maybe we can do this check before wrapping a user text into tags, and replace it automatically, but that probably means searching and updating a lot more places.

File: src/Symfony/Component/Console/Tests/Helper/SymfonyQuestionHelperTest.php [hunk 2]
```diff
@@ -79,7 +79,9 @@ public function testAskReturnsNullIfValidatorAllowsIt()
         $questionHelper = new SymfonyQuestionHelper();
         $questionHelper->setInputStream($this->getInputStream("\n"));
         $question = new Question('What is your favorite superhero?');
-        $question->setValidator(function ($value) { return $value; });
+        $question->setValidator(function ($value) {
+            return $value;
+        });
```
Comment (fabpot) [inline src/Symfony/Component/Console/Tests/Helper/SymfonyQuestionHelperTest.php]: should be reverted. The new PHP-CS-Fixer is not configured the right way, sorry about that.

**Comment by ogizanagi** [comment]
Finally, I think there is only a few places that would require this special check, so it might be worth it and can be fixed within this very same PR IMHO.

@chalasr You can apply https://github.com/symfony/symfony/compare/2.7...ogizanagi:fix/2.7/console/trailing_backslashes if you think it makes sense, or I'll submit another PR once this one is merged.

File: src/Symfony/Component/Console/Tests/Helper/SymfonyQuestionHelperTest.php (hunk 2 above)
Comment (chalasr) [inline src/Symfony/Component/Console/Tests/Helper/SymfonyQuestionHelperTest.php]: good to know, reverted

**Comment by chalasr** [comment]
@ogizanagi It makes good sense to me, thank you for the patch, applied.

File: src/Symfony/Component/Console/Formatter/OutputFormatter.php [hunk 3]
```diff
@@ -33,6 +33,18 @@ public static function escape($text)
...
+    }
+
+    /**
+     * Escapes trailing "\" in given text.
+     *
+     * @param string $text Text to escape
+     *
+     * @return string Escaped text
+     */
+    public static function escapeTrailingBackslash($text)
```
Comment (xabbuh) [inline src/Symfony/Component/Console/Formatter/OutputFormatter.php]: Shouldn't the method be marked as internal?

Comment (ogizanagi) [inline src/Symfony/Component/Console/Formatter/OutputFormatter.php]: Not sure. I guess it can be used by anyone creating it's own helpers in userland.

Comment (fabpot) [inline src/Symfony/Component/Console/Formatter/OutputFormatter.php]: I agree that it should be marked as internal.

Comment (chalasr) [inline src/Symfony/Component/Console/Formatter/OutputFormatter.php]: done

**Comment by chalasr** [comment]
This is ready to go

**Comment by fabpot** [comment]
Thank you @chalasr.

**Pull request merged**